File: common/src/main/java/generations/gg/generations/core/generationscore/network/packets/dialogue/S2CHealDialoguePacket.java
package generations.gg.generations.core.generationscore.network.packets.dialogue;

import com.pokemod.pokemod.network.api.PokeModPacket;
import com.pokemod.pokemod.network.protocol.PokeModClientPacketListener;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.fml.DistExecutor;
import net.minecraftforge.network.NetworkEvent;

import java.util.List;

// TODO delay being able to go to the next node through this packet
public class S2CHealDialoguePacket implements PokeModPacket {

    private final List<String> text;
    private final boolean useNextArrow;

    public S2CHealDialoguePacket(List<String> text, boolean useNextArrow) {
        this.text = text;
        this.useNextArrow = useNextArrow;
    }

    public S2CHealDialoguePacket(FriendlyByteBuf buf) {
        this.text = buf.readList(FriendlyByteBuf::readUtf);
        this.useNextArrow = buf.readBoolean();
    }

    @Override
    public void encode(FriendlyByteBuf buf) {
        buf.writeCollection(text, FriendlyByteBuf::writeUtf);
        buf.writeBoolean(useNextArrow);
    }

    @Override
    public void process(NetworkEvent.Context ctx) {
        DistExecutor.unsafeRunWhenOn(Dist.CLIENT, () -> () ->
                PokeModClientPacketListener.getInstance().handleDialogueSayText(text, useNextArrow));
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/network/packets/dialogue/S2CHealDialoguePacket.kt
package generations.gg.generations.core.generationscore.network.packets.dialogue

import com.cobblemon.mod.common.api.net.ClientNetworkPacketHandler
import com.cobblemon.mod.common.api.net.NetworkPacket
import generations.gg.generations.core.generationscore.GenerationsCore
import generations.gg.generations.core.generationscore.client.screen.dialgoue.display.DialogueScreen
import net.minecraft.client.Minecraft
import net.minecraft.network.FriendlyByteBuf

// TODO delay being able to go to the next node through this packet
class S2CHealDialoguePacket(private val text: List<String>, private val useNextArrow: Boolean) : NetworkPacket<S2CHealDialoguePacket> {
    override val id = ID

    override fun encode(buf: FriendlyByteBuf) {
        buf.writeCollection(text) { obj, value -> obj.writeUtf(value) }
        buf.writeBoolean(useNextArrow)
    }

    companion object {
        val ID = GenerationsCore.id("say_dialogue")

        fun decode(buf : FriendlyByteBuf) = S2CHealDialoguePacket(buf.readList { obj: FriendlyByteBuf -> obj.readUtf() }, buf.readBoolean())
    }

    class Handler: ClientNetworkPacketHandler<S2CHealDialoguePacket> {
        override fun handle(packet: S2CHealDialoguePacket, client: Minecraft) {
            val dialogueScreen = Minecraft.getInstance().screen;
            if (dialogueScreen is DialogueScreen) dialogueScreen.activeInfo = DialogueScreen.SayActiveInfo(packet.text, packet.useNextArrow)
        }
    }
}
